File: backend/app/schemas/skyroute.py
"""
Pydantic Schemas for SkyRoute - Weather-Aware Flight Route Intelligence
"""

from typing import List, Dict, Any, Optional, Literal
from pydantic import BaseModel, Field

class AirportEndpoint(BaseModel):
    icao: str
    iata: str
    name: str
    city: str
    state: str
    latitude: float
    longitude: float
    elevation_ft: Optional[int] = None

class RouteGeometry(BaseModel):
    type: Literal["LineString"] = "LineString"
    coordinates: List[List[float]]  # [[lon, lat], ...]

class CorridorGeometry(BaseModel):
    type: Literal["Polygon"] = "Polygon"
    coordinates: List[List[List[float]]]  # [[[lon, lat], ...]]

class RouteStageInfo(BaseModel):
    title: str
    location_name: str
    risk_level: Literal["LOW", "MODERATE", "HIGH", "SEVERE"]
    weather_condition: str
    temperature_c: float
    visibility_km: float
    wind_kts: float
    summary: str
    concerns: List[str] = []

class WeatherHazardFeature(BaseModel):
    hazard_id: str
    type: Literal["thunderstorm", "rain", "wind", "turbulence", "visibility", "warning"]
    title: str
    severity: Literal["LOW", "MODERATE", "HIGH", "SEVERE"]
    location_description: str
    coordinates_center: List[float]  # [lat, lon]
    geometry: Dict[str, Any]  # GeoJSON Polygon / MultiPolygon / Point
    distance_from_corridor_km: float
    route_impact: Literal["INTERSECTION", "NEAR ROUTE", "OFF ROUTE"]
    expected_time_window: str
    operational_impact: str
    source: str = "IMD"
    updated_at: str
    valid_until: Optional[str] = None
    parameters: Dict[str, Any] = {}

class RouteRiskSummary(BaseModel):
    score: int = Field(ge=0, le=100, description="0=low concern, 100=severe concern")
    level: Literal["LOW", "MODERATE", "HIGH", "SEVERE"]
    confidence: Literal["HIGH", "MODERATE", "LIMITED"]
    available_parameters_count: int
    total_parameters_count: int = 6
    total_hazards: int
    route_intersections: int
    near_route_hazards: int
    major_concern: str
    rationale: str

class SkyRouteData(BaseModel):
    origin: AirportEndpoint
    destination: AirportEndpoint
    distance_nm: int
    distance_km: int
    estimated_flight_time: str
    route_linestring: RouteGeometry
    corridor_polygon: CorridorGeometry
    risk_summary: RouteRiskSummary
    stages: Dict[str, RouteStageInfo]  # departure, enroute, arrival
    hazards: List[WeatherHazardFeature]
    ai_weather_brief: str
    sources: List[Dict[str, str]]
    disclaimer: str
    generated_at: str

class SkyRouteResponse(BaseModel):
    success: bool
    data: Optional[SkyRouteData] = None
    message: Optional[str] = None

File: frontend/src/types/skyroute.ts
/**
 * SkyRoute Frontend Type Definitions
 * Weather-Aware Flight Route Intelligence
 */

export type HazardSeverity = 'LOW' | 'MODERATE' | 'HIGH' | 'SEVERE';
export type HazardType = 'thunderstorm' | 'rain' | 'wind' | 'turbulence' | 'visibility' | 'warning';
export type RouteImpact = 'INTERSECTION' | 'NEAR ROUTE' | 'OFF ROUTE';

export interface AirportEndpoint {
  icao: string;
  iata: string;
  name: string;
  city: string;
  state: string;
  latitude: number;
  longitude: number;
  elevation_ft?: number;
}

export interface RouteGeometry {
  type: 'LineString';
  coordinates: [number, number][]; // [lon, lat]
}

export interface CorridorGeometry {
  type: 'Polygon';
  coordinates: [number, number][][]; // [[lon, lat], ...]
}

export interface RouteStageInfo {
  title: string;
  location_name: string;
  risk_level: HazardSeverity;
  weather_condition: string;
  temperature_c: number;
  visibility_km: number;
  wind_kts: number;
  summary: string;
  concerns: string[];
}

export interface WeatherHazardFeature {
  hazard_id: string;
  type: HazardType;
  title: string;
  severity: HazardSeverity;
  location_description: string;
  coordinates_center: [number, number]; // [lat, lon]
  geometry: {
    type: string;
    coordinates: any;
  };
  distance_from_corridor_km: number;
  route_impact: RouteImpact;
  expected_time_window: string;
  operational_impact: string;
  source: string;
  updated_at: string;
  valid_until?: string;
  parameters: Record<string, any>;
}

export interface RouteRiskSummary {
  score: number;
  level: HazardSeverity;
  confidence: 'HIGH' | 'MODERATE' | 'LIMITED';
  available_parameters_count: number;
  total_parameters_count: number;
  total_hazards: number;
  route_intersections: number;
  near_route_hazards: number;
  major_concern: string;
  rationale: string;
}

export interface SkyRouteData {
  origin: AirportEndpoint;
  destination: AirportEndpoint;
  distance_nm: number;
  distance_km: number;
  estimated_flight_time: string;
  route_linestring: RouteGeometry;
  corridor_polygon: CorridorGeometry;
  risk_summary: RouteRiskSummary;
  stages: {
    departure: RouteStageInfo;
    enroute: RouteStageInfo;
    arrival: RouteStageInfo;
  };
  hazards: WeatherHazardFeature[];
  ai_weather_brief: string;
  sources: { name: string; type: string; updated: string }[];
  disclaimer: string;
  generated_at: string;
}

export interface SkyRouteResponse {
  success: boolean;
  data?: SkyRouteData;
  message?: string;
}
